File: Data Structures & Algorithms/decode-ways/submission-1.cpp
class Solution {
public:
    vector<int>dp;
    int numDecodings(string s) {
        int n = s.length();
        dp = vector<int>(n,-1);
        return solve(s,0);

    }

    int solve(string s, int i){
        if(i >= s.length()) return 1;
        if(s[i] == '0') return 0;
        if(dp[i] != -1) return dp[i];
        if(s[i] > '2'){
            dp[i] =  solve(s,i+1);
        }else if(s[i] == '1'){
            if(i+1 < s.length()){
                dp[i] =  solve(s,i+2) + solve(s,i+1);
            } else{
                dp[i] = solve(s,i+1);
            }
        }else if(s[i] =='2'){
            if(i+1 < s.length() && s[i+1] <= '6'){
                dp[i] = solve(s,i+2) + solve(s,i+1);
            } else{
                dp[i] = solve(s,i+1);
            }
        }
        return dp[i];
    }
    
};

File: Data Structures & Algorithms/subsets-ii/submission-1.java
class Solution {
    Set<List<Integer>> set = new HashSet<>();
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        set.clear();

        Arrays.sort(nums);
        List<Integer> list = new ArrayList<>();

        solve(nums, 0, list);

        List<List<Integer>> ans = new ArrayList<>();

        for (List<Integer> subset : set) {
            ans.add(subset);
        }
        return ans;
    }

    public void solve(int[] nums, int start, List<Integer> curr) {
        int n = nums.length;
        set.add(new ArrayList<>(curr));
        for (int i = start; i < n; i++) {
            curr.add(nums[i]);
            solve(nums, i + 1, curr);
            curr.remove(curr.size() - 1);
        }
    }
}
